### PR TITLE
Add es3.d.ts

### LINKED ISSUE #2410: Create ES3 specific lib.d.ts
By targeting to ES3, the following code (within a class definition):

``` typescript
        setOptionDisable(option: HTMLOptionElement, operation: EnumOperations) {
            ko.applyBindingsToNode(option,
                {
                    disable: BI_OP.has(operation)? this.selectedNumbers().length >= 2 : true
                }, item);
        }
```

compiles, but the `item` variable is not defined. Does really `item` exists in a ES3 standard browser? I wasn't able to find it, so I treated this as a bug.

### LINKED ISSUE #15978: `@lib: ES6` on fourslash test case breaks tests
**TypeScript Version:** master

**Repro:**

1. Insert `// @lib: ES6` on `test/fourslash/commentsFunctionExpression.ts`
2. Run `jake runtests t=commentsFunctionExpression`

**Expected behavior:**

The test should run well

**Actual behavior:**

Numbers unexpectedly becomes `any`:

`AssertionError: At 14: quick info text: expected 'var assigned: (s: string) => any' to equal 'var assigned: (s: string) => number'`

Probably the test code does not correctly check `/// <reference />` comment in lib.d.ts files but not sure how can it be fixed.

### PR DESCRIPTION
**Pull request by saschanaz**
Revives #16077

Fixes #2410
Fixes #15978

Intentionally included only NaN in es3.d.ts to initially generate minimal baseline changes.

### CODE REVIEW

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
+/// <reference lib="es3" />
 /////////////////////////////
 /// ECMAScript APIs
 /////////////////////////////
 
-declare const NaN: number;
 declare const Infinity: number;
```
Comment (ericdrobinson) [inline src/lib/es5.d.ts]: @saschanaz Shouldn't the Infinity constant also be shifted into `es3.d.ts`? See section 4.3.22 of [the spec](https://www.ecma-international.org/publications/files/ECMA-ST-ARCH/ECMA-262,%203rd%20edition,%20December%201999.pdf).

**Comment by ericdrobinson** [comment]
Am I missing something or is the current setup one that would have an [`es3.d.ts`-only](https://github.com/Microsoft/TypeScript/blob/ba3c3ebacfc4fe79b33f70f8a66fe9d7e1c6e4b1/src/lib/es3.d.ts) project have nothing beyond `NaN` defined? Shouldn't the `es3.d.ts` file contain most of what's included in the [current `es5.d.ts`](https://github.com/Microsoft/TypeScript/blob/ba3c3ebacfc4fe79b33f70f8a66fe9d7e1c6e4b1/src/lib/es5.d.ts)?

Unless I'm missing something, the current `es3.d.ts` looks... pretty useless, no?

**Comment by RyanCavanaugh** [comment]
In the intervening years since #2410 was logged, pretty much all the ES3 targets have died out (even the use cases identified most recently don't even really count as a true ES3 target).

Something we should have realized four years ago was that this would be a breaking change! Anyone targeting ES3 but using ES5 APIs (which is an entirely reasonable state to find yourself in if your actual runtime is ES5 and you're not using getters/setters, or if you're using polyfills) is going to be broken if we merge this and make `target: ES3` align to `lib: ES3`.

So given the relative lack of remaining use cases for this, and the fact that to preserve back compat we'd have to make this an opt-in that you can effectively already achieve today with `noLib`, we're going to have to pass on this. Sorry for not identifying that problem earlier - we definitely should have.

**Comment by attilah** [comment]
@RyanCavanaugh FYI Salesforce Marketing Cloud is using an ES3 compatible language set for (Jint is their JS runtime) their SSJS language.

**Comment by ericdrobinson** [comment]
@attilah You might want to mention that over on https://github.com/microsoft/TypeScript/issues/51909 which is explicitly about the ES3 target deprecation coming in TypeScript 5.0 (removal planned for 5.5).

**Pull request closed without merging**